CIP-spendMany-script-purpose (replace Spend)/README.md

```diff
@@ -0,0 +1,116 @@
+---
+CIP: ????
+Title: SpendMany script purpose 
+Category: Plutus
+Status: Proposed
+Authors:
+    - Matteo Coppola <m.coppola.mazzetti@gmail.com>
+Implementors: 
+    - 
+Discussions:
+    - 
+Created: 2024-07-13
+License: Apache-2.0
+---
+
+## Abstract
+
+This CIP proposes to replace the current Spend purpose with another one called SpendMany, removing the need of computational waste and higher fees when a transaction validates multiple inputs coming from the same script.
+
+SpendMany is executed only once for the whole transaction and it comes with 2 parameters:
+- The current script hash
+- The list of the indexes of the inputs that are being spent from this script
```

Comment:
Some scripts allow composing with itself where inputs use *different* redeemers. It sounds like this CIP will require all inputs from the script in this transaction to use the same redeemer.

Reply:
Yes, the current CIP version requires 1 redeemer for all the inputs from the smart contract.
Can't we just use a single redeemer that contains lists for each field, indexed by the order of the inputs?
We already do like that and it works smoothly: as we know the inputs and their order in advance, it's easy to setup such a redeemer.

Reply:
>  Can't we just use a single redeemer that contains lists for each field, indexed by the order of the inputs?

I'd rather not use the redeemer like this since it must always be validated at run-time to verify all of the inputs are properly accounted for. If an input is accidentally/maliciously omitted from the redeemer, the smart contract can return true without actually validating that input. In other words, the security of this approach depends on the DApp developer. It is possible for one DApp to be secure while another DApp is missing checks.

Instead, the ledger can do the checks *before* running the scripts if separate spending groups are used. One of the plutus/ledger devs can correct me if I am wrong, but if the rule is that all inputs in a spending group must use the same redeemer, wouldn't this be a very cheap check that can be part of phase 1 validation? You can just hash the redeemer used for each input in the group and then compare the hashes. If an input is not tagged by a redeemer, the ledger will immediately catch it and fail the transaction.

I have not used the redeemer index pattern very much, so if I am misunderstanding something please correct me.